### PR TITLE
feat: discord bot multitenancy

### PR DESCRIPTION
**Pull request by igorosip0v**
## landing-page

- Make `Install Now` and `Add to Discord` buttons [lead to discord auth page](https://github.com/worldcoin/world-id-discord-bot/pull/10/files#diff-fe33915f7991d05a17d3128a61ad0fbfccd4e6760bea2bda4a000afdb2f43c0eR39), after successful auth it redirects to `/auth` page which gets token from query params and [fetches guilds list](https://github.com/worldcoin/world-id-discord-bot/pull/10/files#diff-9b1656e075cb71dffc545ea440dd06e585146c090d417a35b06493bd93169653R41) from discord API, then passes it and redirect to `/configuration` page.
- Add "Server" field in configuration form with list of guilds where user have Administrator status.
- Add [saving config](https://github.com/worldcoin/world-id-discord-bot/pull/10/files#diff-dc12ff640e699d2575acabf833c304341715d239fed88890bd4b43fd39b6dc4dR75) to DynamoDB.

### Require improvements: 
 - Some data (such as dynamoDB table name, aws creds, discord auth redirect URL) stored in `.env.local` file. May be there is better way to get/store this data.

### How to create discord auth URL:
- Go to [Discord developer portal](https://discord.com/developers).
- Click `New application`
- Go to OAuth2 section -> General.
- Add redirect to a proper url. In my case I've set `https://localhost:3000/auth`
- Go to OAuth2 -> URL Generator
- Set `identity` and `guilds` checkboxes
- Set your saved redirect url
- Get your generated URL
- In generated URL change search param `response_type` value from `code` to `token`. (response_type=token)

## bot updates

- Update context values
    - Remove `action_id`
    - Remove `roles`
    - [Add ](https://github.com/worldcoin/world-id-discord-bot/pull/10/files#diff-bda22a3de7550efc76dd29e2a427a8d5abd5863f6e96a339ff47390d9bed7d14R34)`bot_configs_table_name`
- [Create new DynamoDB table](https://github.com/worldcoin/world-id-discord-bot/pull/10/files#diff-ba79e7ab77e2a8c1afe9047d464f494ea919cd38475f31954d5e9bc5a0b60f34R68) with name from context value.
  - Primary key: `guild_id`: string
  - `action_id` field: string
  - `roles` field: list of strings
- Add readonly permissions to [configure lambda](https://github.com/worldcoin/world-id-discord-bot/pull/10/files#diff-ba79e7ab77e2a8c1afe9047d464f494ea919cd38475f31954d5e9bc5a0b60f34R199), [verification lambda](https://github.com/worldcoin/world-id-discord-bot/pull/10/files#diff-dd1c64018325754f7a07090d92cd139c6f4a8394f718595ec6cfd3055e34205dR170), [role assigning lambda](https://github.com/worldcoin/world-id-discord-bot/pull/10/files#diff-dd1c64018325754f7a07090d92cd139c6f4a8394f718595ec6cfd3055e34205dR129) to read values from DynamoDB.
- Fetch data from DynamoDB to create roles, verify with action_id from DB, assign roles.

### Require improvements: 
- Data fetches from DynamoDB three times: [when bot adds to server](https://github.com/worldcoin/world-id-discord-bot/pull/10/files#diff-8f678d3fbbae5b4f82474487315c76e0204946c051dafccdcd487279f3bed6f8R40), [on verification process](https://github.com/worldcoin/world-id-discord-bot/pull/10/files#diff-8f678d3fbbae5b4f82474487315c76e0204946c051dafccdcd487279f3bed6f8R40), [when roles assigned](https://github.com/worldcoin/world-id-discord-bot/pull/10/files#diff-8f678d3fbbae5b4f82474487315c76e0204946c051dafccdcd487279f3bed6f8R40). I assume there is the way to fetch it once and share it in code, but I'm not sure how to share some data between lambdas.

---
<details>
<summary>🎥 Screencast</summary>

https://user-images.githubusercontent.com/89008845/189528632-658cad92-b15d-42ca-9fb9-0ee8e7cc5858.mp4
</details>

### CODE REVIEW

File: src/landing-page/src/screens/Configuration/index.tsx [hunk 1]
```diff
@@ -131,7 +205,7 @@ export const Configuration = memo(function Configuration() {
               <a
                 href="https://developer.worldcoin.org"
                 target="_blank"
-                rel="noopener"
+                rel="noopener noreferrer"
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/landing-page/src/screens/Configuration/index.tsx:208]: @paolodamico eslint fixes it on save. 

eslint([react/jsx-no-target-blank](https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-no-target-blank.md))

File: src/landing-page/src/screens/Configuration/index.tsx [hunk 2]
```diff
@@ -148,22 +222,51 @@ export const Configuration = memo(function Configuration() {
...
                 className="mt-4"
-                options={discordRolesList}
-                selected={discordRoles}
-                onChange={setDiscordRoles}
+                options={roles}
+                setOptions={setRoles}
+                selected={selectedRoles}
+                onChange={setSelectedRoles}
+                withInput
+                inputPlaceholder="Enter role name"
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/landing-page/src/screens/Configuration/index.tsx:230]: @paolodamico This is not a search. This input allows user to add custom named role to the list and then pick it from the list to use it on the server. You can see how it works on screencast.

File: src/landing-page/src/screens/Configuration/index.tsx [hunk 3]
```diff
@@ -9,26 +9,46 @@ import {
...
 import type { Option } from "./types/option";
 
 const discordRolesList: Array<Option> = [
-  { label: "Moderator", value: "moderator" },
-  { label: "Server admin", value: "server admin" },
-  { label: "Community MVP", value: "community mvp" },
-  { label: "Community MVP 2", value: "community mvp 2" },
-  { label: "Community MVP 3", value: "community mvp 3" },
-  { label: "Community MVP 4", value: "community mvp 4" },
+  { label: "Moderator", value: "Moderator" },
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/landing-page/src/screens/Configuration/index.tsx:21]: @paolodamico Some default role names, so that the user can select them at once

File: src/world-id/resources.queue-processing.ts [hunk 4]
```diff
@@ -219,10 +220,32 @@ export const handler: SQSHandler = async (event, context) => {
...
+  let guildBotConfig;
+
+  try {
+    if (!message.guild_id) {
+      throw Error("guild_id should be provided");
+    }
+    guildBotConfig = await getBotConfig(message.guild_id);
+    console.log("DynamoDB response: ", guildBotConfig);
+  } catch (error) {
+    console.log(error);
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/world-id/resources.queue-processing.ts]: @paolodamico updated

File: src/world-id/resources.queue-processing.ts [hunk 5]
```diff
@@ -125,6 +125,7 @@ async function walletConnectFlow(
 
     // FIXME: Export payload generation function from World ID JS SDK and reuse here
 
+    console.log(`Starting verification with action_id: ${action_id}`);
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/world-id/resources.queue-processing.ts:128]: @paolodamico Let's make a separate task for this. What do you think?

File: src/landing-page/src/screens/Auth/index.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  token: searchParams.get("access_token"),
+};
+
+const isAdmin = (permissions: any) => {
+  // @NOTE https://github.com/discord/discord-api-docs/blob/main/docs/topics/Permissions.md#permissions
+  return (permissions & 0x8) === 0x8;
+};
+
+const getGuilds = async () => {
+  return await fetch("https://discord.com/api/users/@me/guilds", {
```
Comment (gvidon) [inline src/landing-page/src/screens/Auth/index.tsx:22]: @igorosip0v just curious, do they have SDK lib for Discord?

File: src/utils/get-bot-config.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        error: new Error("There is no data for this guild in DB"),
+      };
+    }
+
+    if (data.Item) {
+      const item = unmarshall(data.Item);
+      console.log("DynamoDB response: ", item);
+      result = { data: item as BotConfig, error: undefined };
+    }
+  } catch (error) {
```
Comment (gvidon) [inline src/utils/get-bot-config.ts:51]: @igorosip0v just to avoid nested blocks of code let's only wrap the `client.send()` with try/catch.

File: src/utils/get-bot-config.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  | {
+      data: BotConfig;
+      error?: never;
+    }
+  | {
+      data?: null;
+      error: Error;
+    };
+
+export const getBotConfig = async (guild_id: string) => {
```
Comment (gvidon) [inline src/utils/get-bot-config.ts:21]: @igorosip0v can we explicitly assign a type for the function return value?

File: src/world-id/assign-roles-on-verified.lambda.ts [hunk 9]
```diff
@@ -6,17 +6,25 @@ import type { Handler } from "aws-lambda";
...
 export const handler: Handler<UserCompletedVerification> = async (event) => {
   console.log("User completed verification", event);
 
+  const guildBotConfig = await getBotConfig(event.guild_id);
+
+  if (guildBotConfig.error) {
+    throw Error(`No bot config for guild ${event.guild_id}`);
+  }
+
+  const ROLES_TO_ASSIGN = guildBotConfig.data.roles;
```
Comment (gvidon) [inline src/world-id/assign-roles-on-verified.lambda.ts:26]: @igorosip0v why is it in a cap case?

File: src/landing-page/src/utils/save-bot-config.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+import { marshall } from "@aws-sdk/util-dynamodb";
+
+type BotConfig = {
+  guild_id: string;
+  action_id: string;
+  roles: Array<string>;
+};
+
+// @REVIEW not sure where to store credentials
+const TABLE_NAME = process.env.REACT_APP_AWS_DYNAMODB_TABLE_NAME;
```
[the remark this answers is not in the document]
Comment (gvidon) [reply src/landing-page/src/utils/save-bot-config.ts:13]: @paolodamico @igorosip0v it should be passed from the DB stack down to the landing page stack. But I'm not sure if the landing page already has a stack defined. It's probably not yet deployed anywhere?

File: src/utils/get-bot-config.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
+import { DynamoDBClient, GetItemCommand } from "@aws-sdk/client-dynamodb";
+import { marshall, unmarshall } from "@aws-sdk/util-dynamodb";
+import { getEnv } from "./get-env";
+
+type BotConfig = {
```
Comment (gvidon) [inline src/utils/get-bot-config.ts:5]: @igorosip0v the type repeats the one defined above. Do you think we can move it to common location in the code?

File: src/landing-page/src/utils/dynamodb-client.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
+
+// @REVIEW not sure where to store credentials
+export const client = new DynamoDBClient({
```
[the remark this answers is not in the document]
Comment (gvidon) [reply src/landing-page/src/utils/dynamodb-client.ts:4]: @paolodamico that's correct, we can generate creds in the DB stack, and then pass to landing page stack. But I'm not sure how is the landing page currently deployed.

**Review by gvidon — CHANGES_REQUESTED** [review]
@ottodimi please feel free requesting a review on top of my feedback.

File: src/world-id/resources.ts [hunk 13]
```diff
@@ -29,7 +30,7 @@ export class WorldIdVerifier extends Construct {
     props: {
       defaultLambdaProps?: Partial<NodejsFunctionProps>;
       botToken: ISecret;
-      rolesToAssignToVerifiedUsers: string[];
+      tableName: string;
```
Comment (ottodimi) [inline src/world-id/resources.ts:33]: If you are not provisioning table then take `table: ITable` as parameter (similarly to `botToken: ISecret`).

File: src/world-id/resources.ts [hunk 14]
```diff
@@ -109,16 +110,24 @@ export class WorldIdVerifier extends Construct {
...
     props.botToken.grantRead(rolesAssigningLambda);
 
+    const dynamodbManagedReadonlyPolicy =
+      iam.ManagedPolicy.fromManagedPolicyArn(
+        this,
+        this.node.id,
+        "arn:aws:iam::aws:policy/AmazonDynamoDBReadOnlyAccess",
+      );
+
+    rolesAssigningLambda.role?.addManagedPolicy(dynamodbManagedReadonlyPolicy);
```
Comment (ottodimi) [inline src/world-id/resources.ts:129]: Once you have `table: ITable` use `table.grantRead(rolesAssigningLambda)` instead of granting read only access to all tables in account.

File: src/world-id/assign-roles-on-verified.lambda.ts (hunk 9 above)
Comment (ottodimi) [inline src/world-id/assign-roles-on-verified.lambda.ts:26]: Uppercase is confusing because value is no longer passed through environment variable.

File: src/bot.stack.ts [hunk 15]
```diff
@@ -60,18 +62,44 @@ export class WorldIdDiscordBotStack extends Stack {
...
+        .scaleOnUtilization({ targetUtilizationPercent: 75 });
+
+      table
+        .autoScaleWriteCapacity({
+          minCapacity: 1,
+          maxCapacity: 1200,
+        })
+        .scaleOnUtilization({ targetUtilizationPercent: 75 });
+    } catch (error) {
+      console.log(error);
```
Comment (ottodimi) [inline src/bot.stack.ts:96]: You can't just log error and leave `TABLE_NAME` as empty string.

**Review by ottodimi — CHANGES_REQUESTED** [review]

File: src/bot.stack.ts [hunk 16]
```diff
@@ -60,18 +62,44 @@ export class WorldIdDiscordBotStack extends Stack {
...
-        | undefined) ?? ""
-    )
-      .split(/\s*,\s*/g)
-      .filter((v) => v.trim());
+    let TABLE_NAME = "";
+
+    try {
+      const table = new dynamodb.Table(
+        this,
+        scope.node.tryGetContext("bot_configs_table_name") as string,
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/bot.stack.ts:70]: @gvidon It's the only place where we use this context value. Is it acceptable to just hardcode the table name, or what are the other options?

[the remark this answers is not in the document]
Comment (gvidon) [reply src/bot.stack.ts:70]: @igorosip0v if this is the only place — yeah, let's simply hardcode it. All the consumers of the table within CDK stack will take it from the resource.

File: src/endpoints/oauth2-callback.lambda.ts [hunk 17]
```diff
@@ -33,6 +34,23 @@ export const handler: APIGatewayProxyHandler = async (event) => {
...
+    console.log("DynamoDB response: ", guildBotConfig);
+  } catch (error) {
+    console.log(error);
+  }
+
+  if (!guildBotConfig) {
+    return {
+      statusCode: 422,
+      // @FIXME add correct url for configuration page.
+      body: `Please configure your bot settings first. You can update your configuration here: <Configuration Landing page>`,
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/endpoints/oauth2-callback.lambda.ts:45]: @gvidon There should be a production deployed address with the `/auth` route, something like `https://world-id-discord-bot.com/auth`

File: src/bot.stack.ts [hunk 18]
```diff
@@ -60,18 +62,44 @@ export class WorldIdDiscordBotStack extends Stack {
...
+          },
+        },
+      );
+
+      TABLE_NAME = table.tableName;
+
+      // @REVIEW not sure what is correct values for write/read capacity
+      table
+        .autoScaleReadCapacity({
+          minCapacity: 1,
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/bot.stack.ts]: Updated

File: src/bot.stack.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/bot.stack.ts]: Updated

File: src/bot.stack.ts [hunk 19]
```diff
@@ -60,18 +62,44 @@ export class WorldIdDiscordBotStack extends Stack {
...
+        scope.node.tryGetContext("bot_configs_table_name") as string,
+        {
+          partitionKey: {
+            name: "guild_id",
+            type: dynamodb.AttributeType.STRING,
+          },
+        },
+      );
+
+      TABLE_NAME = table.tableName;
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/bot.stack.ts]: Updated

File: src/bot.stack.ts [hunk 20]
```diff
@@ -60,18 +62,44 @@ export class WorldIdDiscordBotStack extends Stack {
...
+
+      // @REVIEW not sure what is correct values for write/read capacity
+      table
+        .autoScaleReadCapacity({
+          minCapacity: 1,
+          maxCapacity: 600,
+        })
+        .scaleOnUtilization({ targetUtilizationPercent: 75 });
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/bot.stack.ts]: Updated

File: src/utils/get-bot-config.ts (hunk 7 above)
Comment (igorosip0v) [reply src/utils/get-bot-config.ts:37]: Updated

File: src/utils/get-bot-config.ts (hunk 8 above)
Comment (igorosip0v) [reply src/utils/get-bot-config.ts]: Updated

File: src/world-id/assign-roles-on-verified.lambda.ts (hunk 9 above)
Comment (igorosip0v) [reply src/world-id/assign-roles-on-verified.lambda.ts]: Updated

File: src/utils/get-bot-config.ts (hunk 11 above)
Comment (igorosip0v) [reply src/utils/get-bot-config.ts]: Updated

File: src/world-id/resources.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/world-id/resources.ts]: Updated

File: src/world-id/resources.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/world-id/resources.ts]: Updated

File: src/world-id/resources.ts [hunk 21]
```diff
@@ -150,15 +159,15 @@ export class WorldIdVerifier extends Construct {
...
         SIGNAL: this.node.tryGetContext("signal"),
-        ROLES_TO_ASSIGN: props.rolesToAssignToVerifiedUsers.join(","),
         SIGNAL_DESCRIPTION: this.node.tryGetContext("signal_description"),
+        DYNAMODB_TABLE_NAME: props.tableName,
       },
     });
     this.processor.addEventSource(
       new SqsEventSource(this.queue, { batchSize: 1 }),
     );
+    this.processor.role?.addManagedPolicy(dynamodbManagedReadonlyPolicy);
```
[the remark this answers is not in the document]
Comment (igorosip0v) [reply src/world-id/resources.ts]: Updated

File: src/world-id/assign-roles-on-verified.lambda.ts (hunk 9 above)
Comment (igorosip0v) [reply src/world-id/assign-roles-on-verified.lambda.ts]: Updated

File: src/bot.stack.ts (hunk 15 above)
Comment (igorosip0v) [reply src/bot.stack.ts]: @ottodimi Do we need try/catch around table creation? I thought `dynamodb.Table` would throw an error if the table existed, but it seems like it does nothing when the table exists and there are no changes for this table. And if there are some changes,`dynamodb.Table` will apply it to the existing table. What is the cases when table creation can throw an error?

File: src/landing-page/src/utils/dynamodb-client.ts [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+  region: process.env.REACT_APP_AWS_REGION,
+  credentials: {
+    accessKeyId: process.env.REACT_APP_AWS_ACCESS_KEY_ID as string,
+    secretAccessKey: process.env.REACT_APP_AWS_SECRET_ACCESS_KEY as string,
+    sessionToken: process.env.REACT_APP_AWS_SESSION_TOKEN,
```
Comment (ottodimi) [inline src/landing-page/src/utils/dynamodb-client.ts:9]: Usage of `sessionToken` suggests that these are temporary credentials that are going to expire. In general you don't need to specify credentials explicitly (except few special cases when you know what you are doing), just let aws sdk to discover them automatically.

File: src/bot.stack.ts (hunk 15 above)
Comment (ottodimi) [reply src/bot.stack.ts]: Table is not created inside of `try...catch`, it's way more complicated. These constructs (classes that `extends Construct`) declare what infrastructure will look like and how the pieces are connected together. This hierarchy of constructs is then converted into JSON document during `cdk deploy` or `cdk synth` (can be found inside `cdk.out/` directory). JSON documents subsequently are sent to AWS for provisioning. Provisioning (actual process of creating a table) happens entirely in cloud, you cannot catch anything there. You may even reboot your PC and it's still going to continue.

TL;DR: `try...catch` is useless.

Comment (igorosip0v) [reply src/bot.stack.ts]: Got it, thanks!

File: src/landing-page/src/screens/Auth/index.tsx (hunk 6 above)
Comment (igorosip0v) [reply src/landing-page/src/screens/Auth/index.tsx:22]: There is some libs, but it's simplier to make a single request this way in my opinion

**Review by ottodimi — APPROVED** [review]

**Comment by igorosip0v** [comment]
@paolodamico 

I've made a dev branch for a [staging domain](https://staging-wld-id-bot.vercel.app/). It synced with this PR, so you can test updated flow there. Currently staging page working with the bot from the AWS dev account ([configured for staging](https://github.com/worldcoin/world-id-discord-bot/blob/igorosipov/wid-97-discord-bot-multitenancy/src/endpoints/oauth2-callback.lambda.ts#L142)). 

---

https://user-images.githubusercontent.com/89008845/201331849-6719f3a9-8bf8-456f-93fa-fe0ce7b1d293.mp4

**Review by gvidon — APPROVED** [review]
Let's finally merge it!🚀

**Comment by gvidon** [comment]
Closing this as it seems to be no longer relevant. @paolodamico please feel free to revert if I'm wrong.
